src/shellingham/nt.py

```diff
@@ -1,128 +1,144 @@
-        # I think). Note that we need to use 'mbcs' instead of encoding
-        # settings from sys because this is from the Windows API, not Python
-        # internals (which those settings reflect). (pypa/pipenv#3382)
-        executable = pe.szExeFile
-        if isinstance(executable, bytes):
-            executable = executable.decode('mbcs', 'replace')
-
-        info = {'executable': executable}
-        if pe.th32ParentProcessID:
-            info['parent_pid'] = pe.th32ParentProcessID
-        yield pe.th32ProcessID, info
-        success = windll.kernel32.Process32Next(h_process, byref(pe))
-
-
-def _get_executable(process_dict):
+k32.Process32First.argtypes = \
+    (ctypes.wintypes.HANDLE, ctypes.POINTER(ProcessEntry32))
+k32.Process32First.restype = ctypes.wintypes.BOOL
+k32.Process32First.errcheck = _check_expected(ERROR_NO_MORE_FILES)
+
+k32.Process32Next.argtypes = \
+    (ctypes.wintypes.HANDLE, ctypes.POINTER(ProcessEntry32))
+k32.Process32Next.restype = ctypes.wintypes.BOOL
+k32.Process32Next.errcheck = _check_expected(ERROR_NO_MORE_FILES)
+
+k32.GetCurrentProcessId.argtypes = ()
+k32.GetCurrentProcessId.restype = ctypes.wintypes.DWORD
+
+k32.OpenProcess.argtypes = \
+    (ctypes.wintypes.DWORD, ctypes.wintypes.BOOL, ctypes.wintypes.DWORD)
+k32.OpenProcess.restype = ctypes.wintypes.HANDLE
+k32.OpenProcess.errcheck = _check_handle(INVALID_HANDLE_VALUE)
+
+k32.QueryFullProcessImageNameW.argtypes = \
+    (ctypes.wintypes.HANDLE, ctypes.wintypes.DWORD, ctypes.wintypes.LPWSTR,
+     ctypes.wintypes.PDWORD)
+k32.QueryFullProcessImageNameW.restype = ctypes.wintypes.BOOL
+k32.QueryFullProcessImageNameW.errcheck = _check_expected(ERROR_INSUFFICIENT_BUFFER)
+
+
+@contextlib.contextmanager
+def Win32Handle(handle):
     try:
-        executable = process_dict.get('executable')
-    except (AttributeError, TypeError):
-        return None
-    if isinstance(executable, string_types):
-        executable = executable.lower().rsplit('.', 1)[0]
-    return executable
+        yield handle
+    finally:
+        k32.CloseHandle(handle)
+
+
+def enum_processes():
+    with Win32Handle(k32.CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0)) as snap:
+        entry = ProcessEntry32()
+        entry.dwSize = ctypes.sizeof(entry)
+        ret = k32.Process32First(snap, entry)
+        while ret:
+            yield entry
+            ret = k32.Process32Next(snap, entry)
+
+
+def get_full_path(proch):
+    size = ctypes.wintypes.DWORD(ctypes.wintypes.MAX_PATH)
+    while True:
+        path_buff = ctypes.create_unicode_buffer('', size.value)
+        if k32.QueryFullProcessImageNameW(proch, 0, path_buff, size):
+            return path_buff.value
+        size.value *= 2
+
+
+SHELLS = frozenset((
+    b'sh.exe', b'bash.exe', b'dash.exe', b'ash.exe',
+    b'csh.exe', b'tcsh.exe',
+    b'ksh.exe', b'zsh.exe', b'fish.exe',
+    b'cmd.exe', b'powershell.exe', b'pwsh.exe',
+    b'elvish.exe', b'xonsh.exe',
+))
```

Comment:
Can this be calculated from `_core.SHELLS` instead? It’s probably not good to duplicate the list of shells in two places.